Enable TLS support

salieri11 opened the pull request:
This PR finalizes TLS support for the library. It adds self-signed certificates generation scrips and some final cleaning in the codebase.

beerriot commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -326,32 +288,35 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
     }
 
-    string remPeer = sm.str().substr(0, sm.str().length() - 3);
+    string remPeer = sm.str().substr(3, sm.str().length() - 3);
```
Given that it looks like there was a bug here, a comment about what this "3" is would be a nice addition.

salieri11 replied:
Fixed

beerriot commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -750,13 +728,6 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
   }
 
   void write_async(const char *data, uint32_t length) {
-    // async_write(socket,
```
Does this mean that write_async is not actually async?

salieri11 replied:
Yes, I need to change the name. The write operation is sync since we want to have a confidence that the message has been sent before sending another one.

salieri11 replied:
Yes, we want to have more confidence that the message has been sent before sending the next one. Will change the method name.

beerriot commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -811,17 +784,29 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+      << ", to: " << _destId
+      << ", offset: " << offset
+      << ", length: " << length);
+    LOG_TRACE(_logger, "exit, node " << _selfId << ", dest: " << _destId);
```
This trace log is only going to print if the debug log has also printed. It doesn't add any new information. It should probably be removed.

beerriot commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -956,28 +951,36 @@ class TlsTCPCommunication::TlsTcpImpl {
       _pAcceptor = boost::make_unique<tcp::acceptor>(_service, ep);
       start_accept();
-    } else LOG_INFO("skipping listen for node: " << _selfId);
+    } else // clients donn't listen
```
nit: too many n's in "don't"

beerriot commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -956,28 +951,36 @@ class TlsTCPCommunication::TlsTcpImpl {
 
+    // this node should connect only to nodes with lower ID
+    // and all nodes with higher ID will connect to this node
+    // we dont want that clients will connect to another clients
```
nit: "we don't want clients to connection to other clients"

upshaw-alex replied:
nit: Neither "we do**nt** want **that** clients **will** connect to **an**other clients" nor "we don't want clients to connect**ion** to other clients" sounds completely right to me.

beerriot commented on bftengine/tests/simpleTest/replica.cpp:
```diff
@@ -151,6 +153,7 @@ int main(int argc, char **argv) {
   BasicConfigurator config;
   config.configure();
 #endif
+  //std::this_thread::sleep_for(chrono::seconds(20));
```
Please remove dead code.

beerriot commented on CMakeLists.txt:
```diff
@@ -22,19 +20,27 @@ endif()
 endif()
+if(NOT BUILD_COMM_TCP_TLS)
+    set(BUILD_COMM_TCP_TLS FALSE)
+endif()
```
I think we should add a check right after this, to make sure that at most one of these is set, and possibly also to log which communication system was chosen. It's not until the internal build scripts, or even the code, that it's obvious that only one communication system can be chosen at a time, and that it must be chosen at build time, otherwise.

salieri11 replied:
ITs logged in the ConnectionFactory, but of course we can print it from here as well. Someday the comm system will come from configuration - and this flag will be used for build purpose only. But for now you are right, its not clear. WIll add it.

salieri11 replied:
@upshaw-alex , sorry for being late for your comment. In general, the decision of which comm module will be used in the runtime will come from configuration, some day. The purpose of these flags is just to build the correct module. But for now, we use these flag to decide both which module to build and what will be the actual comm module for the runtime.

beerriot commented on CMakeLists.txt:
```diff
@@ -22,19 +20,27 @@ endif()
+
+
+
+
```
nit: this is a lot of added whitespace - is it all needed?

salieri11 replied:
removed

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -247,11 +201,20 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+
+    EC_KEY *ecdh = EC_KEY_new_by_curve_name (NID_secp384r1);
+    if (! ecdh) {
+      LOG_ERROR(_logger, "Unable to create EC");
```
Hmmm... well, in this case I think we should exit with SIGABORT... It means that something bad happened with OpenSSL, probably bad version etc.. And of course we cant ran in this case. Will add exit() where needed.

upshaw-alex commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -326,32 +288,35 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
     try {
-      remotePeerId = stoi(sm.str());
+      remotePeerId = stoi(remPeer);
     } catch (const std::invalid_argument &ia) {
```
Correct me if there is some nuance here I am missing (such as this catch being required for the code to compile, if that is the case), but the fact we are catching an `std::invalid_argument` here for this `stoi` call suggests we do not trust that `remPeer` is a valid string representation of an `int`. If we do distrust the contents of `remPeer`, should we not also be catching an `std::out_of_range`, as I believe `stoi` can also throw one of those?

salieri11 replied:
Thanks, good point. fixed.

upshaw-alex commented on scripts/linux/create_tls_certs.sh:
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env bash
+
+
+#!/usr/bin/env bash
```
I am not really that experienced with shell scripts, so correct me if there is something I am missing, but is there a reason we seem to have this line selecting bash twice?

beerriot commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -299,7 +303,11 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+    // since we use pinning we must know who is the remote peer.
+    // peerNIdPrfixLength stands for the length of 'OU=' substring
+    size_t peerNIdPrfixLength = 3;
+    string remPeer = sm.str().substr(3, sm.str().length() - peerNIdPrfixLength);
```
nit: replace the first `3` here as well

beerriot commented on README.md:
```diff
@@ -168,6 +168,10 @@ use TCP in your application, you need to build the TCP module as mentioned
 and passing PlainTcpConfig object to it.
 
+We also support TCP over TLS communication. To enable it, change the
+`BUILD_COMM_TCP_TLS` flag to `TRUE` in the main CMakeLists.txt file. When
```
Tiny last minute request: add a note in the Dependencies section about installing openssl if using TLS, please.

upshaw-alex commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -971,12 +972,12 @@ class TlsTCPCommunication::TlsTcpImpl {
     // this node should connect only to nodes with lower ID
     // and all nodes with higher ID will connect to this node
-    // we dont want that clients will connect to another clients
+    // we don't want that clients will connect to other clients
```
nit: "we don't want that clients will connect to other clients" still sounds a little off to me somehow; I might suggest "we don't want clients to connect to other clients".

guyg8 commented:
Our TLS module requires Boost. Please update the README. Thanks!





/home/my/concord-bft/bftengine/src/communication/TlsTCPCommunication.cpp:23:10: fatal error: 'boost/bind.hpp' file not found
#include "boost/bind.hpp"
         ^
1 error generated.
bftengine/CMakeFiles/corebft.dir/build.make:1166: recipe for target 'bftengine/CMakeFiles/corebft.dir/src/communication/TlsTCPCommunication.cpp.o' failed
make[2]: *** [bftengine/CMakeFiles/corebft.dir/src/communication/TlsTCPCommunication.cpp.o] Error 1
CMakeFiles/Makefile2:996: recipe for target 'bftengine/CMakeFiles/corebft.dir/all' failed
make[1]: *** [bftengine/CMakeFiles/corebft.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2

beerriot commented on README.md:
```diff
@@ -160,10 +160,29 @@ You may also need to export CPLUS_INCLUDE_PATH variable set to /usr/local/includ
+
+    cd boost_1_64_0
+    ./bootstrap.sh
+    sudo ./b2 --with=system --with=filesystem install
```
Are you sure about these commands, @salieri11? My attempts to use them seem to be ignoring the arguments. I've had better success with:

```
./bootstrap.sh --with-libraries=system,filesystem
./b2
sudo ./b2 install
```

salieri11 replied:
Yes, it worked for me... But your version is official one. I will change the README. Thanks!

guyg8 commented:
In simpleTest, after playing with the connectivity of replica 4, replica 2 crashed because of  segmentation fault:

Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007fc0140000f8 in ?? ()
[Current thread is 1 (Thread 0x7fc018bd5700 (LWP 53379))]
(gdb) bt
#0  0x00007fc0140000f8 in ?? ()
#1  0x00007fc01b569462 in BIO_write () from /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#2  0x0000000000625876 in boost::asio::ssl::detail::engine::put_input (this=0x7fc014015a50, data=...) at /usr/local/include/boost/asio/ssl/detail/impl/engine.ipp:188
#3  0x0000000000625511 in boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> >, boost::asio::ssl::detail::handshake_op, boost::_bi::bind_t<void, boost::_mfi::mf1<void, AsyncTlsConnection, boost::system::error_code const&>, boost::_bi::list2<boost::_bi::value<AsyncTlsConnection*>, boost::arg<1> (*)()> > >::operator() (this=0x7fc018bd4ac8, ec=..., bytes_transferred=0, start=0)
    at /usr/local/include/boost/asio/ssl/detail/io.hpp:240
#4  0x000000000062e044 in boost::asio::detail::binder2<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> >, boost::asio::ssl::detail::handshake_op, boost::_bi::bind_t<void, boost::_mfi::mf1<void, AsyncTlsConnection, boost::system::error_code const&>, boost::_bi::list2<boost::_bi::value<AsyncTlsConnection*>, boost::arg<1> (*)()> > >, boost::system::error_code, unsigned long>::operator() (this=0x7fc018bd4ac8) at /usr/local/include/boost/asio/detail/bind_handler.hpp:127
#5  0x000000000062dfee in boost::asio::asio_handler_invoke<boost::asio::detail::binder2<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> >, boost::asio::ssl::detail::handshake_op, boost::_bi::bind_t<void, boost::_mfi::mf1<void, AsyncTlsConnection, boost::system::error_code const&>, boost::_bi::list2<boost::_bi::value<AsyncTlsConnection*>, boost::arg<1> (*)()> > >, boost::system::error_code, unsigned long> > (function=...) at /usr/local/include/boost/asio/handler_invoke_hook.hpp:69
#6  0x000000000062dfaf in boost_asio_handler_invoke_helpers::invoke<boost::asio::detail::binder2<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> >, boost::asio::ssl::detail::handshake_op, boost::_bi::bind_t<void, boost::_mfi::mf1<void, AsyncTlsConnection, boost::system::error_code const&>, boost::_bi::list2<boost::_bi::value<AsyncTlsConnection*>, boost::arg<1> (*)()> > >, boost::system::error_code, unsigned long>, boost::_bi::bind_t<void, boost::_mfi::mf1<void, AsyncTlsConnection, boost::system::error_code const&>, boost::_bi::list2<boost::_bi::value<AsyncTlsConnection*>, boost::arg<1> (*)()> > > (
    function=..., context=...) at /usr/local/include/boost/asio/detail/handler_invoke_helpers.hpp:37
#7  0x000000000062df4a in boost::asio::ssl::detail::asio_handler_invoke<boost::asio::detail::binder2<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> >, boost::asio::ssl::detail::handshake_op, boost::_bi::bind_t<void, boost::_mfi::mf1<void, AsyncTlsConnection, boost::system::error_code const&>, boost::_bi::list2<boost::_bi::value<AsyncTlsConnection*>, boost::arg<1> (*)()> > >, boost::system::error_code, unsigned long>, boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> >, boost::asio::ssl::detail::handshake_op, boost::_bi::bind_t<void, boost::_mfi::mf1<void, AsyncTlsConnection, boost::system::error_code const&>, boost::_bi::list2<boost::_bi::value<AsyncTlsConnection*>, boost::arg<1> (*)()> > > (function=..., this_handler=0x7fc018bd4ac8)
    at /usr/local/include/boost/asio/ssl/detail/io.hpp:320
#8  0x000000000062dd89 in boost_asio_handler_invoke_helpers::invoke<boost::asio::detail::binder2<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> >, boost::asio::ssl::detail::handshake_op, boost::_bi::bind_t<void, boost::_mfi::mf1<void, AsyncTlsConnection, boost::system::error_code const&>, boost::_bi::list2<boost::_bi::value<AsyncTlsConnection*>, boost::arg<1> (*)()> > >, boost::system::error_code, unsigned long>, boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> >, boost::asio::ssl::detail::handshake_op, boost::_bi::bind_t<void, boost::_mfi::mf1<void, AsyncTlsConnection, boost::system::error_code const&>, boost::_bi::list2<boost::_bi::value<AsyncTlsConnection*>, boost::arg<1> (*)()> > > > (function=..., context=...)
    at /usr/local/include/boost/asio/detail/handler_invoke_helpers.hpp:37
#9  0x000000000062dac9 in boost::asio::detail::reactive_socket_recv_op<boost::asio::mutable_buffers_1, boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> >, boost::asio::ssl::detail::handshake_op, boost::_bi::bind_t<void, boost::_mfi::mf1<void, AsyncTlsConnection, boost::system::error_code const&>, boost::_bi::list2<boost::_bi::value<AsyncTlsConnection*>, boost::arg<1> (*)()> > > >::do_complete (owner=0x1333fa0, base=0x7fc0140022a0) at /usr/local/include/boost/asio/detail/reactive_socket_recv_op.hpp:110
#10 0x000000000061fe2e in boost::asio::detail::task_io_service_operation::complete (this=0x7fc0140022a0, owner=..., ec=..., bytes_transferred=0) at /usr/local/include/boost/asio/detail/task_io_service_operation.hpp:38
#11 0x000000000065ad27 in boost::asio::detail::task_io_service::do_run_one (this=0x1333fa0, lock=..., this_thread=..., ec=...) at /usr/local/include/boost/asio/detail/impl/task_io_service.ipp:372
#12 0x000000000065a736 in boost::asio::detail::task_io_service::run (this=0x1333fa0, ec=...) at /usr/local/include/boost/asio/detail/impl/task_io_service.ipp:149
#13 0x000000000065a0ee in boost::asio::io_service::run (this=0x1331c80) at /usr/local/include/boost/asio/impl/io_service.ipp:59
#14 0x000000000065d866 in std::_Mem_fn_base<unsigned long (boost::asio::io_service::*)(), true>::operator()<, void>(boost::asio::io_service&) const (this=0x13deb88, __object=...)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/functional:583
#15 0x000000000065d780 in std::_Bind<std::_Mem_fn<unsigned long (boost::asio::io_service::*)()> (std::reference_wrapper<boost::asio::io_service>)>::__call<unsigned long, , 0ul>(std::tuple<>&&, std::_Index_tuple<0ul>) (this=0x13deb88, 
    __args=<unknown type in /home/my/builds/concord-bft/debug/bftengine/tests/simpleTest/server, CU 0x79951, DIE 0x101e06>) at /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/functional:1073
#16 0x000000000065d6af in std::_Bind<std::_Mem_fn<unsigned long (boost::asio::io_service::*)()> (std::reference_wrapper<boost::asio::io_service>)>::operator()<, unsigned long>() (this=0x13deb88)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/functional:1131
#17 0x000000000065d5fe in std::_Bind_simple<std::_Bind<std::_Mem_fn<unsigned long (boost::asio::io_service::*)()> (std::reference_wrapper<boost::asio::io_service>)> ()>::_M_invoke<>(std::_Index_tuple<>) (this=0x13deb88)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/functional:1530
#18 0x000000000065d59e in std::_Bind_simple<std::_Bind<std::_Mem_fn<unsigned long (boost::asio::io_service::*)()> (std::reference_wrapper<boost::asio::io_service>)> ()>::operator()() (this=0x13deb88)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/functional:1520
#19 0x000000000065d062 in std::thread::_Impl<std::_Bind_simple<std::_Bind<std::_Mem_fn<unsigned long (boost::asio::io_service::*)()> (std::reference_wrapper<boost::asio::io_service>)> ()> >::_M_run() (this=0x13deb70)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/thread:115
#20 0x00007fc01a604c80 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#21 0x00007fc01a8d56ba in start_thread (arg=0x7fc018bd5700) at pthread_create.c:333
#22 0x00007fc019d6a41d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
 

I suspect it is similar to https://github.com/xtreemfs/xtreemfs/issues/281

andrewjstone commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -558,32 +560,25 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
                                 size_t bytesRead) {
-    // (IG): patch, dont do it, need to fix multithreading
-    LOG_TRACE("enter, node " << _selfId << ", dest: " << _destId);
+    LOG_TRACE(_logger, "enter, node " << _selfId << ", dest: " << _destId);
```
What does this trace tell you? I see `enter, node` or `enter node` in multiple places. Are the function names already traced? If not, perhaps we want to add them, rather than enter node.

andrewjstone commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -651,11 +633,11 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
                                     << (offset + sizeof(_selfId)));
 
-    AsyncTlsConnection::write_async((const char *) _outBuffer,
+    AsyncTlsConnection::write_sync((const char *) _outBuffer,
```
Why change this to write_sync, from write_async?

andrewjstone commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -666,40 +648,45 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+         << ", dest: " << _destId
+         << ", ec: " << ec.message());
     } else {
       if (connected) {
```
Don't `connected` and `_connectedTimer` need to be protected by a lock like they are elsewhere?

andrewjstone commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -749,14 +737,7 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
-    //       boost::asio::placeholders::bytes_transferred));
-
+  void write_sync(const char *data, uint32_t length) {
     if (!connected)
```
Does this need to be protected by a lock?

andrewjstone commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -807,21 +793,32 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
-    LOG_TRACE("enter, node " << _selfId << ", dest: " << _destId);
-
     lock_guard<recursive_mutex> lock(_connectionsGuard);
+    if(_closed)
```
Add a brace or move to a single line please.

andrewjstone commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -1138,7 +1151,6 @@ int TlsTCPCommunication::Stop() {
     return 0;
 
   auto res = _ptrImpl->Stop();
```
nit: `return _ptrImpl->Stop();` rather than using a temporary variable.

salieri11 replied:
I found that it is much easier to debug when you have temp variable.

salieri11 commented on README.md:
```diff
@@ -168,6 +193,10 @@ use TCP in your application, you need to build the TCP module as mentioned
 
+We also support TCP over TLS communication. To enable it, change the
+`BUILD_COMM_TCP_TLS` flag to `TRUE` in the main CMakeLists.txt file. When
+running simpleTest using the testReplicasAndClient.sh - there is no need to create TLS certificates manually. The script will use the `create_tls_certs.sh` (located under the scripts/linux folder) to create certificates. The latter can be used to create TLS files for any number of replicas, e.g. when extending existing tests.
```
yes, this was an assumption we took with @guyg8 - that certificates has been created prior to running the code.

salieri11 commented on README.md:
```diff
@@ -160,6 +160,31 @@ You may also need to export CPLUS_INCLUDE_PATH variable set to /usr/local/includ
+We use Boost both for plain TCP and TLS communication. Please follow these 
+instructions to install this library
+
+Download Boost version 1.64
```
I think building Boost will take a long time. We can discuss it offline.

andrewjstone replied:
Sure. We should have a discussion about this anyway. It's very possible this technique doesn't work for everything. My only goal is to have it easier for a new user to download and build. How we get there doesn't really matter. We can just mark this resolved for now.

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -35,11 +35,13 @@
 #include "openssl/ssl.h"
 #include <openssl/x509.h>
 #include <openssl/x509v3.h>
+#include "Logging.hpp"
```
Will check it. thanks!

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -35,11 +35,13 @@
 using namespace std;
 using namespace bftEngine;
+using namespace concordlogger;
 using namespace boost::asio;
```
will refactor. thanks!

salieri11 replied:
I have read a bit about this question - it appears to be asked a lot. Although google's code style impose not to use it, there are few decent c++ experts that don't see any problems with it. Using namespaces makes the code shorter. Anyway, if there is a need to be explicit on which function you really intend to call, one may always use the fully qualified name. I agree that specifically for Asio it may be a good idea to use fully qualified names, but on the other hand I don't see any reason to write std:: everywhere. Of course, in headers we will never write "using namespace".

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -247,11 +201,24 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+      abort();
+    }
+
+    EC_KEY_free (ecdh);
```
this is safe to do in openssl, they count references. this is the usage example I have found

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -35,11 +35,13 @@
 class AsyncTlsConnection;
 
 using namespace std;
 using namespace bftEngine;
```
yes, this is the correct way to do it. thanks

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -247,11 +201,24 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
-                                  ).string());
+
+    // if we cant create EC DH params, it may mean that SSL version old or
+    // any other crypto related errors, we can't continue with TLS
```
Well the key initialization is an integrated part of setting ssl server. This piece of code doesn't have any existence "rights" without being used specifically in setting up the server. I agree on making short functions, nevertheless, if the code is a part of the function main goal - it may reside there.

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -397,49 +378,53 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
                              << ", connected: " << connected << ", closed: "
                              << _closed);
 
-    lock_guard<recursive_mutex> lock(_connectionsGuard);
```
in the latest code it doesn't exist anymore

salieri11 commented:
@andrewjstone @yuliasherman I have pushed latest changes, implementing the changes from the review we have done.
In essense:
Implemented timeouts for async operations, added more code comments, methods grouped by logics (send, receive, connect..), cipher suite is passed as a parameter. Hope it will be more readable now.

beerriot commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -245,680 +408,790 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
-                            << _closed);
+  void setTimeOut() {
+    _currentTimeout = _currentTimeout == _maxTimeout
+                      ? _minTimeout
```
Shouldn't this be `_maxTimeout`? I'm guessing the intention was to cap the time at max, right?

salieri11 commented on bftengine/tests/simpleTest/test_comm_config.cpp:
```diff
@@ -156,7 +158,7 @@ PlainUdpConfig TestCommConfig::GetUDPConfig(
                  num_of_replicas, config_file_name);
 
-  PlainUdpConfig ret_val(ip, port, buf_length_, nodes, node_id);
+  PlainUdpConfig ret_val(default_listen_ip_, port, buf_length_, nodes, node_id);
```
When running in AWS, the VM's external IP is not available on the VM itself - this is a routing abstraction. In this case you can't call listen on the external IP - it fails. For such environments, we need listen IP address - that defaults to 0.0.0.0 - and the IP addresses from the configuration files are used for connections only. We will add listen IP as a parameter to the configuration file - then the semantics will be "listen IP" and "connection IP".

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -245,418 +219,493 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
   }
 
   void set_tls_client() {
+    _sslContext.set_verify_mode(asio::ssl::verify_peer);
```
For the client mode, this option is igonored
https://www.openssl.org/docs/man1.0.2/man3/SSL_CTX_set_verify.html

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -35,207 +31,185 @@
+  }
 
+  uint32_t get_message_length(const char *buffer,
+                              size_t length) {
```
fixed

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -245,418 +219,493 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+    if (msgLength == 0 || msgLength > _maxMessageLength - 1 - MSG_HEADER_SIZE) {
+      LOG_ERROR(_logger,
+                "on_read_async_header_completed, msgLen: " << msgLength);
+      read_msg_length_async();
```
agree. fixed

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -245,418 +219,493 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+    was_error(ec, "close");
 
-    return (ec != 0);
+    if (_fOnError) {
```
Yeah it's kind of confusing. I usually check the pointers before using but in this case it makes sense to remove this check so it will become clear that the pointer should always be valid.

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -844,81 +968,121 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+    if (_connections.find(id) != _connections.end()) {
+      LOG_ERROR(_logger, "new incoming connection with peer id that already "
+                         "exists, destroying both, peer: " << id);
+      _connections.erase(id);
```
This function is triggered only from the close() method of AsyncTlsConnection, after all timers has been cancelled. When the connection is removed we rely on asio's cleanup via destructors.

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -1044,19 +1220,23 @@ class TlsTCPCommunication::TlsTcpImpl {
   * On success, returns 0.
   */
   int Stop() {
```
Same as for the Start()

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -1027,8 +1202,9 @@ class TlsTCPCommunication::TlsTcpImpl {
 
   int Start() {
-    if (_pIoThread)
+    if (_pIoThread) {
```
In fact Start is called only once, by replica. But to be sure - its a good idea to add mutex.

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -1044,19 +1220,23 @@ class TlsTCPCommunication::TlsTcpImpl {
     _pIoThread->join();
-    _service.reset();
+
+    _connections.clear();
```
For my understanding, calling _service.Stop() aborts all operations, prevents ready handlers to run - and the allows the AsyncTcpConnection object to be destroyed (using shared_from_this) 
after they are removed from the list.
We dont clean sockets explicitly, relying on Asio cleanup (both TLS stream and TCP)

From Asio docs:
" To effect a shutdown, the application will then need to call the io_service object's stop() member function. This will cause the io_service run() call to return as soon as possible, abandoning unfinished operations and without permitting ready handlers to be dispatched."

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -19,223 +31,374 @@
+    _authenticated = value;
+  }
+
+  bool check_replica(NodeNum node) {
```
Oh its really checking the node being replica, and not the client. We dont want to provide status for clients to the application.

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -19,223 +31,374 @@
+
+  uint32_t get_message_length(const char *buffer,
+                              size_t length) {
+    return *(static_cast<const uint32_t *>(
```
Actually I changed it now to be more safe and being explicit about the endianness.

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -19,223 +31,374 @@
+  /**
+   * this method closes the socket for reconnecting to the remote peer
+   */
+  void close_socket() {
```
Changed to clean_for_reconnect and dispose_connection. This is what these functions actually do.

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -19,223 +31,374 @@
+    if (_statusCallback) {
+      bool isReplica = check_replica(_destId);
+      if (isReplica) {
+        PeerConnectivityStatus pcs;
```
bug, thanks!

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -19,223 +31,374 @@
+   */
+  void on_handshake_complete_outbound(const B_ERROR_CODE &ec) {
+    bool err = was_error(ec, "on_handshake_complete_outbound");
+    if (err) {
```
Yes, handle_error is called which disposes the connection

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -19,223 +31,374 @@
+  void on_handshake_complete_inbound(const B_ERROR_CODE &ec) {
+    set_connected(true);
+    bool err = was_error(ec, "on_handshake_complete_inbound");
+    if (err) {
```
Yes, handle_error is called which disposes the connection

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -245,680 +408,790 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
-    // if server has been verified, check that we co
-    if (-1 != expectedPeerId) {
+    // if server has been verified, check peers match
+    if (UNKNOWN_NODE_ID != expectedPeerId) {
```
The check_certificate() function is used both in client and server. When on server, the expectedPeerId will always be set to UNKNOWN - because we don't know who had connected. In this case there is no need to check if the expected peer id is the actual peer id. This is relevant for the client only. Will add comment.

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -245,680 +408,790 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
-      LOG_TRACE("was error, node " << _selfId << ", dest: " << _destId);
+  read_msglength_completed(const B_ERROR_CODE &ec,
+                           const uint32_t bytesRead,
+                           bool first) {
```
The first is used to indicate whether the callback is a result of time bounded operation or not.
The flow is: try to read as much msg length bytes as possible, without starting timer (first = true). If the whole msg length has been received - start read_async for the message (which will complete only after all msg bytes arrive). If only part of the msg length received - start reading the rest of msg length bytes but this time dont allow partial read (you know they are on the way and in this case first = false) - in both cases, timer should be started. The if(!first) is used to cancel the timer that has not been expired yet but the read ALL bytes operation has been completed.

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -245,680 +408,790 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
-    LOG_TRACE("enter, node " << _selfId << ", dest: " << _destId);
-
-    lock_guard<recursive_mutex> lock(_connectionsGuard);
+    auto res = _readTimer.expires_at(boost::posix_time::pos_infin);
```
Well in this case both operations have the same effect - because we have only 1 thread and in this point of time there may be at most 1 pending async_wait operation that need to be cancelled. I think semantically it's better to say "here I don't want to cancel the timer, I just want it not to expire because I have all my data been read" and in close() the timer is cancelled totally.

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -245,680 +408,790 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+        _receiver->onNewMessage(_destId, _inBuffer, bytesRead);
+      }
+    } catch (std::exception &e) {
+      LOG_ERROR(_logger, "read_msg_async_completed, exception:" << e.what());
```
I don't think so. Processing error in the receiver is valid from the communication module point of view.

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -245,680 +408,790 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+   * If the timer tick occurs - shut down the connection.
+   * Probably we need to find the better way to handle timeouts but the
+   * current implementation assumes long enough timeouts to allow to write data
+   * So if the timer occurs before the async write completes - we have a problem
```
Oh, bad phrase. I mean that "we have a problem with this write operation and we are closing the connection" - this is what the code does. The out queue

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -245,680 +408,790 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+  void start_async_write() {
+    asio::async_write(
+        *_socket,
+        asio::buffer(_outQueue.front().data, _outQueue.front().length),
```
From the deque doc:
 std::deque (double-ended queue) is an indexed sequence container that allows fast insertion and deletion at both its beginning and its end. In addition, insertion and deletion at either end of a deque never invalidates pointers or references to the rest of the elements.

I found this flow more intuitive: start async write on the queue front element, and only after it was completed - remove it and move on to the next one.

As for the size() - it is used always under mutex protection.

andrewjstone replied:
You are correct about iterator invalidation and resizing, but this code is still not thread safe. You can't read and write a std::deque from multiple threads without concurrency control. You can be accessing the front of the deque while something is getting pushed on the back. There is no guarantee that the internal data operations are thread safe. length may not be updated correctly, etc...

andrewjstone replied:
@guyg8 Can you weigh in here please?

salieri11 replied:
start_async_write is called from 2 places: async_write_complete and do_write. Both of them locks the mutex before accessing the queue, so the access to size() is always protected.
I understand that the concern is that the front() of the deque is passed to Asio for async write. The write operation only reads from the deque, without modifying it, such that the only concern left is that new elements will be pushed to the deque while async write is in progress on the front element. The removal of the front element and adding new elements are synchronized. I really don't see here any problem.

andrewjstone replied:
Ok, since async_write_complete always happens after that front() is done being used, I believe you are correct. Thanks for the explanation!

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -956,54 +1231,73 @@ class TlsTCPCommunication::TlsTcpImpl {
     }
   }
 
+  void create_outgoing_connection(
```
reconnect() - is the function in AsyncTlsConnection - is called when connection to the peer is timed out and it should be re-connected (this is not the case when the LIVE connection has been broken, then this connection is disposed).
create_outgoing_connection() - is the function in the manager class, that creates outgoing connection object either on the beginning from the Ctor or after existing outgoing connection has been broken and disposed

andrewjstone replied:
Ah, I got it. This will only be called after manager construction when _fOnError fires in the connection class. Then reconnect(), etc.. check that _closed is not set, or they just return, because the manager class is taking over with a new connection. I think part of my confusion was that `_closed` is a bit of a misnomer. The socket actually does get closed in `clean_for_reconnect()`. What the flag is really being used to check is that the socket is being disposed of.

I'd change `_close` to `_disposing` to make this clearer.

Thanks for the explanation. This does make sense to me now.

salieri11 commented on bftengine/src/communication/TlsTCPCommunication.cpp:
```diff
@@ -245,680 +408,790 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+   * @param id
+   * @param conn
+   */
+  void on_connection_authenticated(NodeNum id, ASYNC_CONN_PTR conn) {
```
This method doesnt do any verification, It just adds the authenticated connection to the connections list in the manager class. The actual verification is performed by the AsyncTlsConnection class itself.

andrewjstone replied:
Got it. Thanks!

salieri11 commented:
> Igor, thanks for the changes and explanations! This code is almost ready to go in. I think there are only 2 small things to fix:
> 
> * Change `_closed` to `_disposing` or `_disposed` in the AsyncTlsConnection class for clarity and to differentiate it from `_connected`.
> * Ensure thread safety of the send message deque. I still am fairly certain the way it's being used is unsafe, although I would like Guy to take a look and at least help explain it one way or the other.
> 
> I also still don't quite understand why recursive mutexes are needed in this code at all. Does a single thread ever retake the same lock? It shouldn't in our callback code, and if it does it's likely a bug. If it's needed by boost asio for some reason I'd like to understand exactly why. Recursive mutexes introduce overhead, and tend to obscure the knowledge about where locks start to be taken and when they are actually released for the reader. I proposer replacing them with regular mutexes unless absolutely needed. Maybe we should have a chat with Guy about this also, and get his perspective.
> 
> Once these things are fixed and/or adequately explained I think we'll be ready to merge. Thanks for all the hard work.
 
You are right about the mutex, it's better not to hide possible bugs. I changed it to regular mutex, will run extensive tests to verify.
